src/r1cs_nark_pcd/help_circuit.rs

```diff
@@ -0,0 +1,273 @@
+    HelpField<E>: PrimeField + Absorbable<HelpField<E>>,
+    MainAffine<E>: Absorbable<HelpField<E>>,
+    HelpAffine<E>: Absorbable<MainField<E>>,
+    PC: R1CSNarkPCDConfig<E>,
+    P: PCDPredicate<MainField<E>>,
+{
+    fn generate_constraints(
+        self,
+        cs: ConstraintSystemRef<HelpField<E>>,
+    ) -> Result<(), SynthesisError> {
+        let HelpCircuit {
+            main_avk,
+            main_accumulation_input_instances: main_input_instances,
+            main_old_accumulator_instances,
+            main_new_accumulator_instance,
+            mut main_accumulation_proof,
+            _config_phantom,
+            _predicate_phantom,
+        } = self;
+
+        // Ensure that prior data exist together.
+        assert_eq!(
+            main_input_instances.is_some(),
+            main_old_accumulator_instances.is_some()
+        );
+
+        // Ensure that prior data has the correct length.
+        assert!(
+            main_input_instances.is_none()
+                || main_input_instances.as_ref().unwrap().len() == P::PRIOR_MSG_LEN
+        );
+
+        assert!(
+            main_old_accumulator_instances.is_none()
+                || main_old_accumulator_instances.as_ref().unwrap().len() == P::PRIOR_MSG_LEN
+        );
+
+        /*
+         * Process the inputs
+         */
+
+        let base_case = main_input_instances.is_none();
+        let main_circuit_input_len = MainCircuit::<E, PC, P>::public_input_size();
+        let (main_input_instances, main_old_accumulator_instances) = if base_case {
+            (
+                vec![InputInstance::zero(main_circuit_input_len, MAKE_ZK); P::PRIOR_MSG_LEN],
+                vec![AccumulatorInstance::placeholder(main_circuit_input_len); P::PRIOR_MSG_LEN],
+            )
+        } else {
+            (
+                main_input_instances.unwrap(),
+                main_old_accumulator_instances.unwrap(),
+            )
+        };
+
+        let claimed_input_hash: HelpField<E> =
+            Self::compute_hash(&main_avk, &main_new_accumulator_instance);
+
+        // In the base case, the size of the proof will be incorrect due to the low and high
+        // T commitments in the underlying HP_AS proof. We substitute the base case proof with a
+        // dummy proof to ensure the size of the circuit remains constant. We are able to do so
+        // because the result of the r1cs_nark_as verify does not matter in the base case.
+        if base_case {
+            main_accumulation_proof = r1cs_nark_as::Proof::placeholder(
+                main_circuit_input_len,
+                P::PRIOR_MSG_LEN * 2,
+                MAKE_ZK,
+            );
+        };
+
+        /*
+         * Allocation
+         */
+
+        let claimed_input_hash_var = FpVar::new_input(ns!(cs, "alloc_claimed_input_hash"), || {
+            Ok(claimed_input_hash)
+        })?;
+
+        let base_case_var = Boolean::new_witness(ns!(cs, "alloc_base_case_bit"), || Ok(base_case))?;
```

Comment:
This should be allocated as a input, otherwise the prover can always set it to true, forcing the accumulation part to be ignored.

Reply:
If we allocate base_case_var as an input, the prover in the next recursive step would need to know whether the prior step was a base case for the accumulation right? Currently, I don't think the PCD library gives the prover this information. Would we need to modify the API to include such detail or is there any other way of passing such information around?